cpp/roaring64map.hh

```diff
@@ -794,39 +794,91 @@ public:
+     */
+    void flipClosed(uint64_t min, uint64_t max) {
+        if (min > max) {
           return;
         }
-        uint32_t start_high = highBytes(range_start);
-        uint32_t start_low = lowBytes(range_start);
-        uint32_t end_high = highBytes(range_end);
-        uint32_t end_low = lowBytes(range_end);
+        uint32_t start_high = highBytes(min);
+        uint32_t start_low = lowBytes(min);
+        uint32_t end_high = highBytes(max);
+        uint32_t end_low = lowBytes(max);
 
+        // We put std::numeric_limits<>::max in parentheses to avoid a
+        // clash with the Windows.h header under Windows.
+        const uint32_t uint32_max = (std::numeric_limits<uint32_t>::max)();
+
+        auto start_iter = ensureRangePopulated(start_high, end_high);
+
+        // If start and end land on the same inner bitmap, then we can do the
+        // whole operation in one call.
         if (start_high == end_high) {
-            roarings[start_high].flip(start_low, end_low);
+            auto &bitmap = start_iter->second;
+            bitmap.flipClosed(start_low, end_low);
+            eraseIfEmpty(start_iter);
             return;
         }
-        // we put std::numeric_limits<>::max/min in parentheses
-        // to avoid a clash with the Windows.h header under Windows
-        // flip operates on the range [lower_bound, upper_bound)
-        const uint64_t max_upper_bound =
-            static_cast<uint64_t>((std::numeric_limits<uint32_t>::max)()) + 1;
-        roarings[start_high].flip(start_low, max_upper_bound);
-        roarings[start_high++].setCopyOnWrite(copyOnWrite);
-
-        for (; start_high <= highBytes(range_end) - 1; ++start_high) {
-            roarings[start_high].flip((std::numeric_limits<uint32_t>::min)(),
-                                      max_upper_bound);
-            roarings[start_high].setCopyOnWrite(copyOnWrite);
+
+        // Because start and end don't land on the same inner bitmap,
+        // we need to do this in multiple steps:
+        // 1. Partially flip the first bitmap in the closed interval
+        //    [start_low, maxUint32]
+        // 2. Flip intermediate bitmaps completely: [0, maxUint32]
+        // 3. Partially flip the last bitmap in the closed interval
+        //    [0, end_low]
+
+        auto num_intermediate_bitmaps = end_high - start_high - 1;
+
+        // 1. Partially flip the first bitmap...
+        {
+            auto &bitmap = start_iter->second;
+            bitmap.flipClosed(start_low, uint32_max);
+            auto temp = start_iter++;
+            eraseIfEmpty(temp);
         }
 
-        roarings[start_high].flip((std::numeric_limits<uint32_t>::min)(),
-                                  end_low);
-        roarings[start_high].setCopyOnWrite(copyOnWrite);
+        // 2. Flip intermediate bitmaps completely...
+        for (uint32_t i = 0; i != num_intermediate_bitmaps; ++i) {
+            auto &bitmap = start_iter->second;
```

Comment:
Optional: Could we add a small comment here which mentions why we can assume all bitmaps in the range exist? Something like:
```
// We can directly use the iterator for the entire range, because we made sure it was populated above.
```

Reply:
I added a comment above the call to `ensureRangePopulated()`, making it consistent with the corresponding comment in `addRangeClosed()` pending in https://github.com/RoaringBitmap/CRoaring/pull/397. Let me know if you think this is good enough.